File: destack-py/destack/graph/memory/graph.py
from collections.abc import Collection, Sequence
from datetime import datetime
from typing import TYPE_CHECKING, Optional, override

from destack.language import EMPTY_LIST, Entity, Graph, Node, NodeType, TraitType, expand_node_types
from destack.language.registry import NODE_CLASS_BY_TYPE
from destack.utils.fractional import INTEGER_ZERO
from destack.utils.uuid import UUID

if TYPE_CHECKING:
    from destack.language import Event, NodeType, Snapshot, TraitType

type_ = type


class MemoryGraph(Graph):
    """A Graph that stores Nodes in memory."""

    __slots__ = ("nodes_by_id", "nodes_by_parent")

    def __init__(self):
        self.nodes_by_id: dict[UUID, Entity] = {}
        self.nodes_by_parent: dict[UUID, dict[NodeType, list[Entity]]] = {}

    #
    # Meta
    #

    @override
    async def open(self) -> None:
        pass

    @override
    async def close(self) -> None:
        pass

    #
    # Write
    #

    def snapshot(
        self,
        space_id: UUID,
        branch_id: UUID | None,
        snapshot_id: UUID | None,
        epoch: int,
    ) -> "Snapshot":
        """Create a Snapshot."""
        raise NotImplementedError

    def insert(
        self,
        space_id: UUID,
        branch_id: UUID | None,
        snapshot_id: UUID | None,
        entities: "Sequence[Entity]",
    ) -> None:
        """Insert Entities into the Graph directly."""
        raise NotImplementedError

    @override
    def append(self, events: "Sequence[Event]") -> None:
        """Append Events to the Graph. EditEvents are reflected immediately."""
        raise NotImplementedError

    @override
    def restate(self, events: "Sequence[Event]") -> None:
        """Restate Events to the Graph. EditEvents are reflected immediately."""
        raise NotImplementedError

    @override
    async def prune(
        self,
        space_id: UUID,
        branch_id: UUID | None,
        snapshot_id: UUID | None,
    ) -> None:
        """Prune the Graph."""
        raise NotImplementedError

    @override
    async def commit(self) -> None:
        """Ensure Events/Entities are persisted in the Graph."""
        raise NotImplementedError

    #
    # Read
    #

    @override
    def seek(
        self,
        space_id: UUID,
        branch_id: UUID | None,
        snapshot_id: UUID | None,
        type: NodeType | Collection[NodeType] | None = None,
        after: datetime | int | None = None,
        before: datetime | int | None = None,
    ) -> "Sequence[Event]":
        """Seek Events from the Graph."""
        raise NotImplementedError

    @override
    def get(
        self,
        id: UUID,
        space_id: UUID,
        branch_id: UUID,
        snapshot_id: UUID,
        include_deleted: bool = False,
    ) -> Optional["Entity"]:
        return self.nodes_by_id.get(id)

    @override
    def get_children[M: "Entity" = "Entity"](
        self,
        node: "Node",
        space_id: UUID,
        branch_id: UUID,
        snapshot_id: UUID,
        type: NodeType | type[M] | None = None,
        include_deleted: bool = False,
    ) -> Sequence[M]:
        # bail if no children
        if not self.nodes_by_parent:
            return ()
        children_by_type = self.nodes_by_parent.get(node.id)
        if not children_by_type:
            return ()

        if type is None:
            # collect children across all types
            children: list = []
            is_ordered = False
            for children_of_type in children_by_type.values():
                node_cls = type_(children_of_type[0])
                if TraitType.ORDERED in node_cls.__traits__:
                    is_ordered = True
                children.extend(children_of_type)
            if is_ordered:
                children.sort(key=lambda n: getattr(n, "order_key", INTEGER_ZERO))
            return children
        else:
            # turn into type
            node_types = expand_node_types(type, expand_inheritance=True)
            if not node_types:
                return ()
            node_cls = NODE_CLASS_BY_TYPE[node_types[0]]

            # collect
            if len(node_types) == 1:
                # collect for single node type
                children: list = children_by_type.get(node_types[0], EMPTY_LIST)
                if children and TraitType.ORDERED in node_cls.__traits__:
                    children.sort(key=lambda n: getattr(n, "order_key", INTEGER_ZERO))
                return children
            else:
                # collect for trait (multiple node types)
                children: list = []
                for node_type in node_types:
                    children.extend(children_by_type.get(node_type, EMPTY_LIST))
                if children and TraitType.ORDERED in node_cls.__traits__:
                    children.sort(key=lambda n: getattr(n, "order_key", INTEGER_ZERO))
                return children

    @override
    def get_ancestors[M: "Entity" = "Entity"](
        self,
        node: "Entity",
        space_id: UUID,
        branch_id: UUID,
        snapshot_id: UUID,
        type: NodeType | type[M] | None = None,
        include_deleted: bool = False,
    ) -> Sequence[M]:
        ancestors: list[Entity] = []
        current = node.parent_ptr
        node_types = expand_node_types(type, expand_inheritance=True)

        # traverse up the parent chain
        while current is not None:
            parent_node = self.get(current.id, space_id, branch_id, snapshot_id)
            if parent_node is None:
                break
            if node_types is None or parent_node.metatype in node_types:
                ancestors.append(parent_node)
            current = parent_node.parent_ptr

        return ancestors  # type: ignore (must be right type)

    @override
    def get_descendants[M: "Entity" = "Entity"](
        self,
        node: "Entity",
        space_id: UUID,
        branch_id: UUID,
        snapshot_id: UUID,
        type: NodeType | type[M] | None = None,
        include_deleted: bool = False,
    ) -> Sequence[M]:
        if not self.nodes_by_parent:
            return ()
        queue: list[Entity] = [node]
        descendants: list[Entity] = []
        node_types = expand_node_types(type, expand_inheritance=True)

        # BFS
        while queue:
            current = queue.pop(0)
            children_by_type = self.nodes_by_parent.get(current.id)
            if not children_by_type:
                continue
            for children_of_type in children_by_type.values():
                queue.extend(children_of_type)

            if not node_types:
                for children_of_type in children_by_type.values():
                    descendants.extend(children_of_type)
            else:
                for node_t in node_types:
                    descendants.extend(children_by_type.get(node_t, ()))

        return descendants  # type: ignore (must be right type)

File: destack-ts/src/graph/memory/graph.ts
import type { Entity } from "@destack/language/core/builtin";
import { NodeType, TraitType } from "@destack/language/core/builtin/common";
import { expandNodeTypes, Graph } from "@destack/language/core/runtime/graph";
import { NODE_CLASS_BY_TYPE } from "@destack/language/registry";
import { INTEGER_ZERO } from "@destack/utils/fractional";

/** A Graph that stores Entities in memory. */
export class MemoryGraph extends Graph {
  private nodesById: Map<string, Entity>;
  private nodesByParent: Map<string, Map<NodeType, Entity[]>>;

  constructor() {
    super();
    this.nodesById = new Map();
    this.nodesByParent = new Map();
  }

  override get(id: string): Entity | null {
    return this.nodesById.get(id) ?? null;
  }

  override has(id: string): boolean {
    return this.nodesById.has(id);
  }

  override clear(): void {
    // nodes
    this.nodesById.clear();
    this.nodesByParent.clear();
  }

  override add(node: Entity): void {
    const existing = this.nodesById.get(node.id);
    if (existing !== undefined) {
      throw new Error(`node ${node.repr()} already in ${this.repr()}: ${existing.repr()}`);
    }
    // node
    this.nodesById.set(node.id, node);
    // parent
    const parentPtr = (node as any).parentPtr;
    if (parentPtr !== null && parentPtr !== undefined) {
      if (!this.nodesByParent.has(parentPtr.id)) {
        this.nodesByParent.set(parentPtr.id, new Map());
      }
      const childNodeType = node.metatype;
      const parentMap = this.nodesByParent.get(parentPtr.id)!;
      if (!parentMap.has(childNodeType)) {
        parentMap.set(childNodeType, []);
      }
      parentMap.get(childNodeType)!.push(node);
    }
  }

  override remove(node: Entity): void {
    // parent
    const parentPtr = (node as any).parentPtr;
    if (parentPtr !== null && parentPtr !== undefined) {
      const parentMap = this.nodesByParent.get(parentPtr.id);
      if (parentMap) {
        const childNodeType = node.metatype;
        const children = parentMap.get(childNodeType);
        if (children) {
          const index = children.indexOf(node);
          if (index !== -1) {
            children.splice(index, 1);
          }
          if (children.length === 0) {
            parentMap.delete(childNodeType);
            if (parentMap.size === 0) {
              this.nodesByParent.delete(parentPtr.id);
            }
          }
        }
      }
    }
    // node
    this.nodesById.delete(node.id);
  }

  override getChildren(options: { node: Entity; nodeType?: NodeType }): Entity[] {
    // bail if no children
    if (this.nodesByParent.size === 0) {
      return [];
    }
    const childrenByType = this.nodesByParent.get(options.node.id);
    if (!childrenByType) {
      return [];
    }

    if (options.nodeType === undefined) {
      // collect children across all types
      const children: Entity[] = [];
      let isOrdered = false;
      for (const childrenOfType of childrenByType.values()) {
        if (childrenOfType.length > 0) {
          const nodeClass = NODE_CLASS_BY_TYPE[childrenOfType[0].metatype];
          if (nodeClass.__definition__.traits.includes(TraitType.ORDERED)) {
            isOrdered = true;
          }
          children.push(...childrenOfType);
        }
      }
      if (isOrdered) {
        children.sort((a, b) => {
          const aOrderKey = (a as any).orderKey ?? INTEGER_ZERO;
          const bOrderKey = (b as any).orderKey ?? INTEGER_ZERO;
          return aOrderKey.localeCompare(bOrderKey);
        });
      }
      return children;
    } else {
      // turn into type
      const nodeTypes = expandNodeTypes(options.nodeType, { expandInheritance: true });
      if (!nodeTypes || nodeTypes.length === 0) {
        return [];
      }
      const nodeClass = NODE_CLASS_BY_TYPE[nodeTypes[0]];

      // collect
      if (nodeTypes.length === 1) {
        // collect for single node type
        const children = childrenByType.get(nodeTypes[0]) ?? [];
        if (children.length > 0 && nodeClass.__definition__.traits.includes(TraitType.ORDERED)) {
          const sortedChildren = [...children];
          sortedChildren.sort((a, b) => {
            const aOrderKey = (a as any).orderKey ?? INTEGER_ZERO;
            const bOrderKey = (b as any).orderKey ?? INTEGER_ZERO;
            return aOrderKey.localeCompare(bOrderKey);
          });
          return sortedChildren;
        }
        return children;
      } else {
        // collect for trait (multiple node types)
        const children: Entity[] = [];
        for (const nodeType of nodeTypes) {
          children.push(...(childrenByType.get(nodeType) ?? []));
        }
        if (children.length > 0 && nodeClass.__definition__.traits.includes(TraitType.ORDERED)) {
          children.sort((a, b) => {
            const aOrderKey = (a as any).orderKey ?? INTEGER_ZERO;
            const bOrderKey = (b as any).orderKey ?? INTEGER_ZERO;
            return aOrderKey.localeCompare(bOrderKey);
          });
        }
        return children;
      }
    }
  }

  override getDescendants(options: { node: Entity; nodeType?: NodeType }): Entity[] {
    if (this.nodesByParent.size === 0) {
      return [];
    }

    const queue: Entity[] = [options.node];
    const descendants: Entity[] = [];
    const nodeTypes = expandNodeTypes(options.nodeType, { expandInheritance: true });

    // BFS
    while (queue.length > 0) {
      const current = queue.shift()!;
      const childrenByType = this.nodesByParent.get(current.id);
      if (!childrenByType) {
        continue;
      }

      // add all children to queue
      for (const childrenOfType of childrenByType.values()) {
        queue.push(...childrenOfType);
      }

      // collect descendants based on type filter
      if (!nodeTypes) {
        for (const childrenOfType of childrenByType.values()) {
          descendants.push(...childrenOfType);
        }
      } else {
        for (const nodeType of nodeTypes) {
          descendants.push(...(childrenByType.get(nodeType) ?? []));
        }
      }
    }

    return descendants;
  }

  /**
   * Get ancestors of a node (going up the parent chain).
   * If a type is specified, only ancestors of that type are collected.
   */
  getAncestors(node: Entity, nodeType?: NodeType): Entity[] {
    const ancestors: Entity[] = [];
    let currentPtr = (node as any).parentPtr;
    const nodeTypes = expandNodeTypes(nodeType, { expandInheritance: true });

    // traverse up the parent chain
    while (currentPtr !== null && currentPtr !== undefined) {
      const parentNode = this.get(currentPtr.id);
      if (parentNode === null) {
        break;
      }
      if (nodeTypes === null || nodeTypes.includes(parentNode.metatype)) {
        ancestors.push(parentNode);
      }
      currentPtr = (parentNode as any).parentPtr;
    }

    return ancestors;
  }
}
